### PR TITLE
Add manual branch coverage for translate

### LINKED ISSUE #13: Manual branch coverage for translate

### PR DESCRIPTION
**Pull request by filippanilsson**
Fix #13 by adding flags and extra testing to display results

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 1]
```diff
@@ -97,19 +98,27 @@ public boolean isSet(final OPTION option) {
...
+                if(firstChar == 'X'){flags[5] = true;}
+
                 start++;
                 isHex = true;
 
                 // Check there's more than just an x after the &#
                 if (start == seqEnd) {
+                    flags[6] = true;
                     return 0;
                 }
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java:125]: Add a flag for the else

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 2]
```diff
@@ -119,39 +128,66 @@ public int translate(final CharSequence input, final int index, final Writer out
             while (end < seqEnd && ( input.charAt(end) >= '0' && input.charAt(end) <= '9' ||
                                     input.charAt(end) >= 'a' && input.charAt(end) <= 'f' ||
                                     input.charAt(end) >= 'A' && input.charAt(end) <= 'F' ) ) {
+                flags[7] = true; // for the while
+                flags[8] = true; // for the &&
+                flags[9] = true; // three inner &&
+                flags[10] = true;
+                flags[11] = true;
+                flags[12] = true; // two inner ||
+                flags[13] = true;
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java]: Should distinguish between the or branches if using separate flags

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 3]
```diff
@@ -119,39 +128,66 @@ public int translate(final CharSequence input, final int index, final Writer out
...
 
             final boolean semiNext = end != seqEnd && input.charAt(end) == ';';
+            flags[14] = true; // && above
 
             if (!semiNext) {
+                flags[15] = true;
                 if (isSet(OPTION.semiColonRequired)) {
+                    flags[16] = true;
                     return 0;
                 }
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java:156]: Add a flag for the else

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 4]
```diff
@@ -119,39 +128,66 @@ public int translate(final CharSequence input, final int index, final Writer out
...
                 if (isSet(OPTION.errorIfNoSemiColon)) {
+                    flags[17] = true;
                     throw new IllegalArgumentException("Semi-colon required at end of numeric entity");
                 }
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java]: Add a flag for the else

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 5]
```diff
@@ -119,39 +128,66 @@ public int translate(final CharSequence input, final int index, final Writer out
...
             try {
                 if (isHex) {
+                    flags[18] = true;
                     entityValue = Integer.parseInt(input.subSequence(start, end).toString(), 16);
                 } else {
+                    flags[19] = true;
                     entityValue = Integer.parseInt(input.subSequence(start, end).toString(), 10);
                 }
             } catch (final NumberFormatException nfe) {
                 return 0;
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java:174]: maybe add a flag here?

Comment (filippanilsson) [reply src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java:174]: Where? Do you mean the catch statement?

Comment (annamaartensson) [reply src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java:174]: Yes

File: src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java [hunk 6]
```diff
@@ -57,6 +57,7 @@ public enum OPTION {
 
     // TODO?: Create an OptionsSet class to hide some of the conditional logic below
     private final EnumSet<OPTION> options;
+    public boolean[] flags = new boolean[26];
```
Comment (annamaartensson) [inline src/main/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaper.java]: Update this to reflect the correct size

File: src/test/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaperTest.java [hunk 7]
```diff
@@ -52,26 +52,46 @@ public void testSupplementaryUnescaping() {
...
+        for (Boolean b : neu.flags) {
+            if (b) {
+                numBranchVisited++;
+            } else {
+                System.out.println("Branch nr: " + i + " never taken");
+            }
+            i++;
+        }
+        System.out.println("Number of branches visited: " + numBranchVisited);
+        System.out.println("Branch Coverage: " + numBranchVisited/26.0000);
```
Comment (annamaartensson) [inline src/test/java/org/apache/commons/lang3/text/translate/NumericEntityUnescaperTest.java]: Same thing here

**Review by annamaartensson — APPROVED** [review]
LGTM

**Review by Tian0602 — APPROVED** [review]
LGTM

**Pull request merged**